Modules/mathmodule.c

```diff
@@ -3306,9 +3306,87 @@ math_comb_impl(PyObject *module, PyObject *n, PyObject *k)
 }
 
 
+/*[clinic input]
+math._as_integer_ratio
+    x: object
+    /
+
+Return integer ratio.
+
+Return a pair of integers, whose ratio is exactly equal to the original
+number and with a positive denominator.
+[clinic start generated code]*/
+
+static PyObject *
+math__as_integer_ratio(PyObject *module, PyObject *x)
+/*[clinic end generated code: output=2e4f43d93f6e7850 input=b54b48dd6bbe22ea]*/
+{
+    _Py_IDENTIFIER(as_integer_ratio);
+    _Py_IDENTIFIER(numerator);
+    _Py_IDENTIFIER(denominator);
+    PyObject *ratio, *as_integer_ratio, *numerator, *denominator;
+
+    if (PyLong_CheckExact(x)) {
+        return PyTuple_Pack(2, x, _PyLong_One);
+    }
+
+    if (_PyObject_LookupAttrId(x, &PyId_as_integer_ratio, &as_integer_ratio) < 0) {
```

Comment:
Based on the context, I can roughly tell what this conditional is doing. From my understanding, ``_PyObject_LookupAttrId()`` is assessing whether or not the PyObject *x* contains an ``as_integer_ratio`` attribute. If a value less than zero is returned (usually -1), it does not contain that attribute. 

However, I'm not certain that I understand where ``PyId_as_integer_ratio`` is coming from or how ``PyId`` actually works. I was unable to find any documentation on ``PyId`` or ``_Py_IDENTIFIER``, so I'm guessing it's an internal part of the C-API (since it's prefixed with an underscore). 

My best guess is that a reference to ``PyId_as_integer_ratio`` was created when ``_Py_IDENTIFIER(as_integer_ratio)`` was used.

I'm fairly new to the C-API, so I'm trying to learn more about it so that I can be more helpful in PR reviews that involve it. Particularly the internal implementation details that aren't in the documentation.

Reply:
It is described in the header: `Include/cpython/object.h`.

Reply:
Ah, thanks for letting me know where to look. The code comments there addressed my question:

> PyId_foo is a static variable, either on block level or file level. On first usage, the string "foo" is interned, and the structures are linked.